#166840920 User Authentication with OAuth2.0

tolumide-ng opened the pull request:
#### What does this PR do?
Enables user signup with google, facebook and gtihub


#### Description of Task to be completed?
- Make a request to **localhost/<port>/auth/github** and approve to signup with Github
- Make a request to **localhost/auth/<port>/facebook** and approve to signup with Facebook
- Make a request to **localhost/auth/<port>/google** and approve to signup with google


#### How should this be manually tested?
Clone the repo, cd into it and run **npm run start:dev**, makes a request to **localhost:<PORT>/auth/google**, and choose the email address you want to signup with


#### Any background context you want to provide?
Provide all details required as specified in .env.example


#### What are the relevant pivotal tracker stories?
#166840920

Onyimatics reviewed (approved):
Nice work @tolumide-ng. Looks good to me.

segunolalive commented on src/helpers/passport/callback.js:
```diff
@@ -0,0 +1,26 @@
+  const { displayName, emails, photos } = payload;
+  const email = emails[0].value;
+  try {
+    const confirmUser = await db.User.findOrCreate({
```
Please use the BaseRepository instead of calling the user model directly

segunolalive replied:
Please use the BaseRepository.

segunolalive commented on src/helpers/passport/facebook.js:
```diff
@@ -0,0 +1,28 @@
+});
+
+passport.deserializeUser((id, done) => {
+  db.User.findById(id).then(user => done(null, user));
```
Avoid calling the models directly

segunolalive commented on src/index.js:
```diff
@@ -4,10 +4,27 @@ import cors from 'cors';
 const app = express();
+
+app.use(
+  cookieSession({
```
why are we using cookies/sessions

segunolalive commented on src/test/oauth.spec.js:
```diff
@@ -0,0 +1,73 @@
+      .reply(200, { message: 'hit the route' });
+
+    const res = await chai.request(app).get('/auth/facebook');
+    expect(res.body.message).to.include('hit the route');
```
why are we sending this message to users? 🤷‍♂ 

I don't think it provides any business value

segunolalive commented on src/test/oauth.spec.js:
```diff
@@ -0,0 +1,73 @@
+      .reply(200, { message: 'hit the route' });
+
+    const res = await chai.request(app).get('/auth/github/callback');
+    expect(res.body).to.an('object');
```
`res.body` is always an object. That's what `bodyParser` does. This doesn't test that your code functions as it should. Remove it and test what that your route returns the appropriate data

segunolalive reviewed (commented):
👍  @tolumide-ng 

Check out the feedback

tolumide-ng committed:
feat(users): Enable OAuth for users (social authentication)

- adds google passport strategy
- adds github passport strategy
- adds facebook passport strategy
- adds .env to .eslintignore, .prettierignore and .gitgnore files
- user can signup with github
- user can signup with google
- user can signup with facebook
- modifies config file

[Delivers #166840920]

timi-codes commented on .coveralls.yml:
```diff
@@ -1 +1,3 @@
 repo_token: COVERALLS_REPO_TOKEN
+exclude_patterns:
```
Thanks for adding this.

timi-codes commented on src/controllers/user.controller.js:
```diff
@@ -145,7 +142,7 @@ class UserController {
    */
   static async verifyUser(req, res) {
-    const { email } = req.token;
+    const { email } = req.currentUser;
```
Thanks for renaming

timi-codes commented on src/database/config/config.js:
```diff
@@ -1,12 +1,12 @@
-const config = require('dotenv');
```
Is there any reason why you chose to rename this?

tolumide-ng replied:
Yes @timi-codes , I did this for consistency. There are other parts in our codebase that import `dotenv` and use it this way

timi-codes commented on src/database/config/config.js:
```diff
@@ -19,10 +19,10 @@ module.exports = {
+    username: process.env.DB_TEST_USERNAME,
+    password: process.env.DB_TEST_PASSWORD || '',
+    database: process.env.DB_TEST_DATABASE,
+    host: process.env.DB_TEST_HOST || '127.0.0.1',
```
I think there is no reason doing '|| 127.0.0.1' localhost will always be used to run your test.

timi-codes reviewed (changes requested):
This appears good to me. However, I left some comments
